Add `parse-outer` command to pyk

gtrepta opened the pull request:
This starts implementing a command with pyk that handles the outer parsing of a K definition into a KAST JSON that can be passed to inner parsing in the java frontend. So far it can handle requires clauses of other files, do markdown selection, and handle extra include paths passed in from the command line.

gtrepta committed:
Add `pyk parse-outer` cli command

gtrepta committed:
Add tests

Baltoli commented on pyk/src/pyk/__main__.py:
```diff
@@ -389,6 +393,60 @@ def exec_json_to_kore(options: JsonToKoreOptions) -> None:
+            check_dir_path(include_path)
+        except ValueError:
+            _LOGGER.critical(f"Could not find directory '{include}' passed to -I")
+            exit(1)
```
Worth noting this isn't what `-I` does in the existing compiler frontend, nor in GCC / clang. Could maybe be a lower level warning rather than an error eventually?

gtrepta replied:
Yep, I can just make this a warning.

gtrepta committed:
Add pyupgrade code changes

gtrepta committed:
Turn bad -I parameter into just a warning

gtrepta committed:
_ast_to_kast

gtrepta committed:
Convert the Definition and output JSON

gtrepta committed:
Add --main-module option

gtrepta committed:
Update TODO

gtrepta committed:
Remove definitions with unsupported ast->kast items from the tests

tothtamas28 commented on pyk/src/tests/integration/test_pyk.py:
This should be a unit test.

tothtamas28 commented on pyk/src/pyk/__main__.py:
```diff
@@ -389,6 +395,65 @@ def exec_json_to_kore(options: JsonToKoreOptions) -> None:
     sys.stdout.write('\n')
 
 
+def exec_parse_outer(options: ParseOuterOptions) -> None:
```
Please extract a function into one of the `pyk.kast` modules that performs the transformation logic. Unit tests can call that function directly.

gtrepta replied:
Sorry, what transformation logic are you referring to?

tothtamas28 replied:
This function performs side effects. You can extract a function that returns `result_text` (or something more structured), that way, the command line handler becomes very simple.

gtrepta replied:
If you mean `_ast_to_kast`, I agree I should write some unit tests for that.

tothtamas28 replied:
Sounds good.

You've already set up a test (`TestParseOuter`) though, what I'm saying, if you extract a function (that takes `main_file`, etc. and returns the text you'd otherwise print), then you can call that directly without stubbing `sys.argv`.

Baltoli replied:
Yes, I agree that the substantive work here should be extracted into a function that _returns_ something - the implementation will look pretty similar to what you have here, just with a different signature

tothtamas28 commented on pyk/src/pyk/konvert/_ast_to_kast.py:
```diff
@@ -0,0 +1,55 @@
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+@singledispatch
```
Nice.

Baltoli replied:
This is a trick I didn't know about; very cool

gtrepta replied:
I had to massage the type checker a bit to accept this in a few places, but it does make for some convenient calls like the one here:

https://github.com/runtimeverification/k/blob/7cbf5551b5aca2e7a19a231611c8ceeb08bc5f0a/pyk/src/pyk/konvert/_ast_to_kast.py#L35

Baltoli replied:
I guess it saves you a big pile of `isinstance` checks which is nice

gtrepta replied:
Came up with a better solution than the KSentence cast above: a10e68aeed9ce8fd4d3cfbb7cfa59bb1089980e8

Baltoli commented on pyk/src/pyk/__main__.py:
```diff
@@ -394,6 +396,12 @@ def exec_json_to_kore(options: JsonToKoreOptions) -> None:
 
 
 def exec_parse_outer(options: ParseOuterOptions) -> None:
+    # TODO Desugar K language syntax to KAST compatible items
```
Perhaps turn this into a tracking issue and assign yourself to it

gtrepta committed:
Remove TODO. Opened issue #4278

Baltoli commented on pyk/src/pyk/__main__.py:
```diff
@@ -389,6 +395,65 @@ def exec_json_to_kore(options: JsonToKoreOptions) -> None:
+    #      Checks:
+    #        - Circular imports
+    #        - Modules with duplicate names
+    main_file_dir = Path(options.main_file.name).resolve().parent
```
Do you want to `check_file_path` this, or is the error from `.resolve()` good enough?

gtrepta replied:
`options.main_file` is checked by the argument parser, so before we get to this `.resolve()` it should error there:

https://github.com/runtimeverification/k/blob/7cbf5551b5aca2e7a19a231611c8ceeb08bc5f0a/pyk/src/pyk/__main__.py#L602

Baltoli replied:
Nice!

gtrepta committed:
Add unit tests for _ast_to_kast

gtrepta committed:
Move outer parsing logic into kast module, so __main__ just handles CLI
arguments

gtrepta committed:
Add unit test for parse_outer

gtrepta committed:
Revert "Add tests"

This reverts commit 39fc2e5d6edf4ac0bdcc5f9c1bd7c53724143e66.

gtrepta committed:
No default lookup directory, and print lookups on error (match kompile
behavior)

gtrepta committed:
Merge branch 'develop' into pyk/parse-outer

tothtamas28 commented on pyk/src/pyk/kast/utils.py:
```diff
@@ -0,0 +1,61 @@
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+def _slurp(
```
Consider sliding this below `parse_outer`.

tothtamas28 commented on pyk/src/pyk/konvert/_ast_to_kast.py:
This module belongs to `pyk.kast` I think.

tothtamas28 commented on pyk/src/pyk/kast/utils.py:
```diff
@@ -0,0 +1,61 @@
+
+    modules = _slurp(text, search_paths, (), md_selector)
+    final_definition = _ast_to_kast(Definition(modules), main_module=main_module)
+    return cast('KDefinition', final_definition)
```
`cast` does not do any runtime checking, consider asserting the type instead.

tothtamas28 commented on pyk/src/pyk/kast/utils.py:
```diff
@@ -0,0 +1,61 @@
+        except ValueError:
+            _LOGGER.critical(f'{require.path} not found')  # TODO Print the source location of the requires clause
+            _LOGGER.critical(f'Lookup directories: {[str(path) for path in search_paths]}')
+            exit(1)
```
`exit` should not be called here since we might want to call `parse_outer` from contexts other than the CLI. Instead, consider re-raising the exception and catching it in `__main__` to do the `exit`.

tothtamas28 commented on pyk/src/tests/unit/test_konvert.py:
```diff
@@ -49,3 +55,40 @@ def test_unmunge(symbol: str, expected: str) -> None:
     assert actual == expected
+
+
+AST_TO_KAST_TEST_DATA: Final = [
```
These tests can go to their own test module, under `tests/unit/kast`.

gtrepta committed:
Replace static type cast with assertion at runtime

gtrepta committed:
Fix possible double reading of required files.

Any files processed by nested slurps need to be propogated up to the
processed_files variable of the slurp that called it. I'm choosing to do this
with a mutable list.

gtrepta committed:
Fix AttributeError when main checks for a config

gtrepta committed:
Call exit from main instead of the utility function

gtrepta committed:
Resolve the main definition file and include reading it in the logs

gtrepta committed:
Move _slurp below parse_outer

gtrepta committed:
Move _ast_to_kast to the kast module

gtrepta committed:
Make a new generic _sentence_to_ksentence method to remove type cast

tothtamas28 commented on pyk/src/pyk/kast/_ast_to_kast.py:
```diff
@@ -0,0 +1,62 @@
+
+@singledispatch
+@_ast_to_kast.register
+def _sentence_to_ksentence(s: Sentence) -> KSentence:
```
This is a placeholder for later implementation, right?

Regardless, it is not safe to trick the type checker this way. Instead consider a helper function

```python3
def _transform_sentences(sentences: Iterable[Sentence]) -> list[KSentence]:
    return []
```

which you can call in `_module_to_kflatmodule`.

gtrepta replied:
It's a placeholder in the same sense that `_ast_to_kast` is, which is that it isn't fully implemented. I can implement sentence transformations with `_sentence_to_ksentence.register`, and I think the dispatcher in `_ast_to_kast` should pass any `Sentence`s along to here.

I'm not sure what you mean by tricking the type checker. This is the same pattern as declaring `_import_to_kimport` and using it in `_module_to_kflatmodule`.

tothtamas28 replied:
> I'm not sure what you mean by tricking the type checker.

This:

```python3
def _sentence_to_ksentence(s: Sentence) -> KSentence:
    return None # type: ignore
```

Because here:

```python3
sentences = (_sentence_to_ksentence(s) for s in m.sentences)
```

you'll end up with a bunch of `None`s, whereas you'd expect `KSentence` instances.

So consider either weakening the signature to:

```python3
def _sentence_to_ksentence(s: Sentence) -> KSentence | None:
    return None
```

and filtering out `None`-s on the call site, or postponing the `singledispatch`-based implementation until all the cases are covered. Then, you can implement this signature as

```python3
def _sentence_to_ksentence(s: Sentence) -> KSentence:
    raise AssertionError(f'Unimplemented conversion for: {type(s)}')
```

gtrepta replied:
Ok, I see what you are saying. Well, there really is no use case for returning `None` other than as a placeholder, so I can change it to raise an `AssertionError` which wont affect anything as there aren't any tests that hit it and it isn't being used in production. It should also make it helpful as I implement the conversions and add tests for them, since the raised exception will be easier to spot than an error message from the logger.

tothtamas28 commented on pyk/src/tests/unit/kast/test_utils.py:
```diff
@@ -0,0 +1,33 @@
+    actual = parse_outer(test_file, main_module, (test_file.parent, test_file.parent / 'include'))
+
+    # Then
+    assert actual.to_json() == expected.to_json()
```
Does this work?

```suggestion
    assert actual == expected
```

gtrepta committed:
Skip json conversion when comparing kast structures

gtrepta committed:
Raise an exception instead of returning None for unimplemented ast->kast
conversions

rv-jenkins committed:
Merge branch 'develop' into pyk/parse-outer
